File: lib/parser.test.ts
import { test, expect } from 'bun:test';

import { parse, length, StyleKind } from './parser.ts';

test('parse empty string', () => {
  expect(parse('')).toEqual([]);
});

test('parse text', () => {
  expect(parse('hello')).toEqual([{ type: 'text', text: 'hello' }]);
});

test('parse newline', () => {
  expect(parse('hello\nworld')).toEqual([
    { type: 'text', text: 'hello' },
    { type: 'newline' },
    { type: 'text', text: 'world' }
  ]);
});

test('parse basic tags', () => {
  expect(parse('[b]hello[/b] world')).toEqual([
    { type: 'style', style: StyleKind.BOLD, enable: true },
    { type: 'text', text: 'hello' },
    { type: 'style', style: StyleKind.BOLD, enable: false },
    { type: 'text', text: ' world' }
  ]);
});

test('parse nested tags', () => {
  expect(parse('[b]hello [i]world[/i][/b]')).toEqual([
    { type: 'style', style: StyleKind.BOLD, enable: true },
    { type: 'text', text: 'hello ' },
    { type: 'style', style: StyleKind.ITALIC, enable: true },
    { type: 'text', text: 'world' },
    { type: 'style', style: StyleKind.ITALIC, enable: false },
    { type: 'style', style: StyleKind.BOLD, enable: false }
  ]);
});

test('parse color', () => {
  expect(parse('[color=#014554]colored text![/color]')).toEqual([
    { type: 'color', color: '#014554', enable: true },
    { type: 'text', text: 'colored text!' },
    { type: 'color', enable: false }
  ]);
});

test('parse escaped tags', () => {
  expect(parse('\\[[b]hello\\[/b]')).toEqual([
    { type: 'escape' },
    { type: 'text', text: '[' },
    { type: 'style', style: StyleKind.BOLD, enable: true },
    { type: 'text', text: 'hello' },
    { type: 'escape' },
    { type: 'text', text: '[/b]' }
  ]);
});

test('parse incorrectly nested tags and escapes', () => {
  expect(parse('now [b[url]https://[i]example.com[/url] is \\ wrong here \\ [/i] \\')).toEqual([
    { type: 'text', text: 'now [b' },
    { type: 'style', style: StyleKind.LINK, enable: true },
    { type: 'text', text: 'https://' },
    { type: 'style', style: StyleKind.ITALIC, enable: true },
    { type: 'text', text: 'example.com' },
    { type: 'style', style: StyleKind.LINK, enable: false },
    { type: 'text', text: ' is ' },
    { type: 'escape' },
    { type: 'text', text: ' wrong here ' },
    { type: 'escape' },
    { type: 'text', text: ' ' },
    { type: 'style', style: StyleKind.ITALIC, enable: false },
    { type: 'text', text: ' ' },
    { type: 'escape' }
  ]);
});

test('parse a bunch of fake tags', () => {
  expect(parse('these [tags] are invalid ]')).toEqual([
    { type: 'text', text: 'these [tags] are invalid ]' }
  ]);
  expect(parse('[url]]teehee[/color ] yea [] ]')).toEqual([
    { type: 'style', style: StyleKind.LINK, enable: true },
    { type: 'text', text: ']teehee[/color ] yea [] ]' }
  ]);
});

test('length', () => {
  expect(length(parse('hello'))).toBe(5);
  expect(length(parse('hello\nworld'))).toBe(11);
});

test('length with styles', () => {
  expect(length(parse('hello [b]world[/b]'))).toBe(11);
  expect(length(parse('[i]goodnight [b]world[/b]'))).toBe(15);
});

test('length with emotes', () => {
  expect(length(parse('hello world [:D] !'))).toBe(15);
});

test('length with emojis', () => {
  expect(length(parse('🤔☃'))).toBe(2);
  expect(length(parse('this is a fox 🦊 from canada 🇨🇦'))).toBe(30);
});

File: src/parser_test.rs

#[cfg(test)]
mod tests {

  use crate::parser::*;

  #[test]
  fn test_parse_empty_string() {
    assert_eq!(parse(""), vec![]);
  }

  #[test]
  fn test_parse_text() {
    assert_eq!(parse("hello"), vec![Part::Text("hello".to_string())]);
  }

  #[test]
  fn test_parse_newline() {
    assert_eq!(
      parse("hello\nworld"),
      vec![
        Part::Text("hello".to_string()),
        Part::Newline,
        Part::Text("world".to_string())
      ]
    );
  }

  #[test]
  fn test_parse_basic_tags() {
    assert_eq!(
      parse("[b]hello[/b] world"),
      vec![
        Part::Style(StyleKind::Bold, true),
        Part::Text("hello".to_string()),
        Part::Style(StyleKind::Bold, false),
        Part::Text(" world".to_string())
      ]
    );
  }

  #[test]
  fn test_parse_nested_tags() {
    assert_eq!(
      parse("[b]hello [i]world[/i][/b]"),
      vec![
        Part::Style(StyleKind::Bold, true),
        Part::Text("hello ".to_string()),
        Part::Style(StyleKind::Italic, true),
        Part::Text("world".to_string()),
        Part::Style(StyleKind::Italic, false),
        Part::Style(StyleKind::Bold, false)
      ]
    );
  }

  #[test]
  fn test_parse_color() {
    let color = Color::new(168, 52, 207);
    assert_eq!(
      parse("[color=#a834cf]colored text![/color]"),
      vec![
        Part::Color(color, true),
        Part::Text("colored text!".to_string()),
        Part::Color(Color::default(), false)
      ]
    );
  }

  #[test]
  fn test_parse_escaped_tags() {
    assert_eq!(
      parse("\\[[b]hello\\[/b]"),
      vec![
        Part::Escape,
        Part::Text("[".to_string()),
        Part::Style(StyleKind::Bold, true),
        Part::Text("hello".to_string()),
        Part::Escape,
        Part::Text("[/b]".to_string())
      ]
    );
  }

  #[test]
  fn test_parse_incorrectly_nested_tags_and_escapes() {
    assert_eq!(
      parse("now [b[url]https://[i]example.com[/url] is \\ wrong here \\ [/i] \\"),
      vec![
        Part::Text("now [b".to_string()),
        Part::Style(StyleKind::Link, true),
        Part::Text("https://".to_string()),
        Part::Style(StyleKind::Italic, true),
        Part::Text("example.com".to_string()),
        Part::Style(StyleKind::Link, false),
        Part::Text(" is ".to_string()),
        Part::Escape,
        Part::Text(" wrong here ".to_string()),
        Part::Escape,
        Part::Text(" ".to_string()),
        Part::Style(StyleKind::Italic, false),
        Part::Text(" ".to_string()),
        Part::Escape
      ]
    );
  }

  #[test]
  fn test_parse_a_bunch_of_fake_tags() {
    assert_eq!(
      parse("these [tags] are invalid ]"),
      vec![Part::Text("these [tags] are invalid ]".to_string())]
    );
    assert_eq!(
      parse("[url]]teehee[/color ] yea [] ]"),
      vec![
        Part::Style(StyleKind::Link, true),
        Part::Text("]teehee[/color ] yea [] ]".to_string())
      ]
    );
  }

  #[test]
  fn test_length_normal_text() {
    assert_eq!(length(&parse("hello")), 5);
    assert_eq!(length(&parse("hello\nworld")), 11);
  }

  #[test]
  fn test_length_with_styles() {
    assert_eq!(length(&parse("hello [b]world[/b]")), 11);
    assert_eq!(length(&parse("[i]goodnight [b]world[/b]")), 15);
  }

  #[test]
  fn test_length_with_emotes() {
    assert_eq!(length(&parse("hello world [:D] !")), 15);
  }

  #[test]
  fn test_length_with_emojis() {
    assert_eq!(length(&parse("🤔☃")), 2);
    assert_eq!(length(&parse("this is a fox 🦊 from canada 🇨🇦")), 30);
  }

}
